### PR TITLE
Moved cartesian transformation code to CartesianRepresentation

### PR DESCRIPTION
**Pull request by astrofrog**
I found that for another project (WCSAxes) I wanted to use the representation classes and wanted to be able to apply arbitrary transforms - the code to do this was already in astropy but not in a place where it could easily be used, so this adds a new ``transform`` method to ``CartesianRepresentation``.

@eteq - I'd like to include this in 1.2 if possible, could you review and let me know what you think?

### CODE REVIEW

**Comment by pllim** [comment]
Would be nice to have an example, either in docstring or user-facing documentation.

**Comment by astrofrog** [comment]
@pllim - done, this adds an example to the docstring which then also appears in the API docs.

**Comment by mhvk** [comment]
@astrofrog: two things: I think that using `tensordot` one can avoid the reshaping operations; also, neither `tensordot` nor `dot` always preserve quantities correctly. How about the following:
```
xyz = self.xyz
old_vec = xyz.value
matrix_value = getattr(matrix, 'value', matrix)
new_vec = np.tensordot(matrix_value, old_vec, ((1,), (0,)))
new_unit = xyz.unit * getattr(matrix, 'unit', u.dimensionless_unscaled)
new_vec = Quantity(new_vec, new_unit, copy=False)
return self.__class__(*new_vec)
```
I do think this needs some test cases...

p.s  Eventually we may want to define `__matmul__` and `__rmatmul__`...

**Comment by mhvk** [comment]
Actually, why didn't I check immediately: I see `tensordot` just hides the reshaping.  So perhaps the best option is just to be explicit:
```
xyz = self.xyz
# Convert matrices to regular arrays.
if isinstance(matrix, np.matrix):
    matrix = np.array(matrix)
new_vec = np.sum(np.rollaxis(matrix, -2, matrix.ndim),
                 np.rollaxis(xyz, 0, xyz.ndim)[..., np.newaxis, :],
                 axis=-1)
return self.__class__(*(np.rollaxis(new_vec, -1, 0)))
```
This is simple enough that one might hope it would be hard to optimize!

**Comment by mhvk** [comment]
Actually, a bit less complicated still is:
```
new_vec = np.sum(matrix, np.rollaxis(xyz, 0, xyz.ndim)[..., np.newaxis], axis=-2)
```
Advantage of this all is that in principle one can pass in a whole stack of matrices, and those will get broadcast correctly.

**Comment by astrofrog** [comment]
@mhvk - just to check, did you mean a ``*`` instead of your first comma?

**Comment by astrofrog** [comment]
@mhvk - I think it should be:

```
vec_new = np.sum(matrix.T * np.rollaxis(xyz, 0, xyz.ndim)[..., np.newaxis], axis=-2)
```

(you were missing a transpose and ``*``).

**Comment by astrofrog** [comment]
@mhvk - just for info, for large arrays, the one liner ends up being slower than this PR, but for small arrays it's 20 times faster.

**Comment by astrofrog** [comment]
While less satisfying, the following code is faster for smaller arrays and the same speed as this PR for large arrays:

```python
    tmp = np.zeros_like(x.value)
    xn = np.zeros_like(x.value)
    yn = np.zeros_like(y.value)
    zn = np.zeros_like(z.value)
    
    for row, out_array in enumerate([xn, yn, zn]):
    
        for column, in_array in enumerate([x.value, y.value, z.value]):
    
            tmp[:] = in_array[:]
            np.multiply(tmp, matrix[row, column], out=tmp)
            np.add(out_array, tmp, out=out_array)
```

In any case, I think the optimal solution would in fact be (for this and other calculations in representation) to have some cython functions to do the heavy lifting of the math. I think for these classes performance is much more important than code length.

**Comment by astrofrog** [comment]
In any case, I would like for us to merge the PR with the same algorithm as before, and look into improvements to the code after the 1.2 feature freeze.

File: astropy/coordinates/representation.py [hunk 1]
```diff
@@ -290,6 +290,64 @@ def from_cartesian(cls, other):
...
+        xyz = self.xyz
+
+        # Since the underlying data can be n-dimensional, reshape to a
+        # 2-dimensional (3, N) array.
+        vec = xyz.reshape((3, xyz.size // 3))
+
+        # Do the transformation
+        vec_new = np.dot(np.asarray(matrix), vec)
+
+        # Reshape to preserve the original shape
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: I see your point with keeping the same algorithm. But maybe at least
```
vec_new = vec_new.reshape(xyz.shape)
return self.__class__(*vec_new)
```

Comment (astrofrog) [inline astropy/coordinates/representation.py]: That does provide improvements for both large and small arrays, so 👍

**Comment by eteq** [comment]
Whoa, definitely a good idea!  I'd thought about exposing the `astropy.coordinates.utils.cartrepr_from_matmul`, but it always seemed awkward - this is a way better idea!

**Comment by eteq** [comment]
Oh, but one other consideration, @astrofrog: the `astropy.coordinates.utils.cartrepr_from_matmul` function is more complicated because it allows broadcasting of Nx3x3 matricies onto length-N cartesian representations (This is critical for some use cases where e.g. frame attributes are length N).  I think your function here doesn't do that, right?  Do you have any objection to making it do so?  Would be easy enough to just pull the code straight from `cartrepr_from_matmul`.

So the idea would be that ``matrix`` could either be 3x3 (which would be the same as this) *or* Nx3x3 (which would do what `cartrepr_from_matmul` does).  Another option would be to add a second method `transform_many` or similar, but I think a single method is cleaner...

**Comment by astrofrog** [comment]
@eteq - I think it would be easier to think about this in a separate PR (and it would strictly only *add* functionality to CartesianRepresentation.transform) since I won't have time to do that this week. One thing I'm not sure about is that CartesianRepresentation doesn't have to contain just 1-dimensional data, so if it contains MxN data, then do we want to accept an MxNx3x3 matrix?

In short, I would like us to merge this before the feature freeze if possible, and I'm happy to work later on generalizing it.

File: astropy/coordinates/tests/test_representation.py [hunk 2]
```diff
@@ -656,6 +656,19 @@ def test_getitem_scalar(self):
...
+
+        matrix = np.array([[1,2,3], [4,5,6], [7,8,9]])
+
+        s2 = s1.transform(matrix)
+
+        assert_allclose(s2.x.value, [1 * 1 + 2 * 3 + 3 * 5, 1 * 2 + 2 * 4 + 3 * 6])
+        assert_allclose(s2.y.value, [4 * 1 + 5 * 3 + 6 * 5, 4 * 2 + 5 * 4 + 6 * 6])
+        assert_allclose(s2.z.value, [7 * 1 + 8 * 3 + 9 * 5, 7 * 2 + 8 * 4 + 9 * 6])
+
+
```
Comment (mhvk) [inline astropy/coordinates/tests/test_representation.py]: Could you add a check that the unit is still OK? `np.dot` usually removes units, but by what I think is happenchance doesn't here, at least for numpy 1.11. I just want to be sure it is OK on all numpy versions.

**Comment by mhvk** [comment]
@astrofrog - see minor comment on tests. With that, I think it is fine to merge and raise an issue to remind us to move the multidimensional matrix to here as well (for that, you probably need one of the solutions we discussed above, so they won't go to waste!).

**Comment by eteq** [comment]
@astrofrog - OK, I see your point.  My initial concern was the need to get this right now because it sets the API, but now that you said that I realize that making this more like `cartrepr_from_matmul` would still be backwards-compatible anyway.  So Agreed that it makes sense to merge this first and add that in later.

**Comment by eteq** [comment]
So, just to be clear, I'm also 👍  on this once you've had a chance to respond to @mhvk's final suggestion here.

**Comment by astrofrog** [comment]
@mhvk @eteq - I've added the test of the units, and opened a new issue to keep track of the broadcasting: https://github.com/astropy/astropy/issues/4878

**Pull request merged**